File: packages/application/src/process-overdue-invoices.use-case.ts
import type { IInvoiceRepository, ILogger, Invoice } from '@monolegal/domain';
import { InvoiceStatus, startOfDay } from '@monolegal/shared';

export interface ProcessOverdueInvoicesResult {
  transitioned: number;
  failed: number;
}

export class ProcessOverdueInvoicesUseCase {
  constructor(
    private readonly invoiceRepository: IInvoiceRepository,
    private readonly logger: ILogger,
  ) {}

  async execute(referenceDate: Date = new Date()): Promise<ProcessOverdueInvoicesResult> {
    const dueDateBefore = startOfDay(referenceDate);
    const invoices = await this.invoiceRepository.findByStatusAndDueDateBefore(
      InvoiceStatus.AL_DIA,
      dueDateBefore,
    );

    this.logger.info('Starting overdue invoice processing', {
      totalInvoices: invoices.length,
      referenceDate: dueDateBefore.toISOString(),
    });

    let transitioned = 0;
    let failed = 0;

    for (const invoice of invoices) {
      try {
        await this.processInvoice(invoice, referenceDate);
        transitioned++;
      } catch (error) {
        failed++;
        this.logger.error('Failed to transition overdue invoice', {
          invoiceId: invoice.id,
          clientId: invoice.clientId,
          status: invoice.status,
          dueDate: invoice.dueDate.toISOString(),
          error: error instanceof Error ? error.message : String(error),
        });
      }
    }

    this.logger.info('Overdue invoice processing completed', { transitioned, failed });

    return { transitioned, failed };
  }

  private async processInvoice(invoice: Invoice, referenceDate: Date): Promise<void> {
    if (!invoice.shouldTransitionToFirstReminder(referenceDate)) {
      return;
    }

    const nextStatus = invoice.getStatusAfterBecomingOverdue();
    await this.invoiceRepository.updateStatus(invoice.id, nextStatus);

    this.logger.info('Overdue invoice transitioned to first reminder', {
      invoiceId: invoice.id,
      clientId: invoice.clientId,
      previousStatus: invoice.status,
      newStatus: nextStatus,
      dueDate: invoice.dueDate.toISOString(),
    });
  }
}

File: backend/dotnet/src/Monolegal.Application/ProcessOverdueInvoicesUseCase.cs
using Monolegal.Domain.Entities;
using Monolegal.Domain.Ports;
using Monolegal.Shared;

namespace Monolegal.Application;

public record ProcessOverdueInvoicesResult(int Transitioned, int Failed);

public class ProcessOverdueInvoicesUseCase(IInvoiceRepository invoiceRepository, ILogger logger)
{
    public async Task<ProcessOverdueInvoicesResult> ExecuteAsync(DateTime? referenceDate = null, CancellationToken cancellationToken = default)
    {
        var reference = referenceDate ?? DateTime.Now;
        var dueDateBefore = DummyInvoiceData.StartOfDay(reference);
        var invoices = await invoiceRepository.FindByStatusAndDueDateBeforeAsync(
            InvoiceStatus.AlDia, dueDateBefore, cancellationToken);

        logger.Info("Starting overdue invoice processing", new Dictionary<string, object?>
        {
            ["totalInvoices"] = invoices.Count,
            ["referenceDate"] = dueDateBefore.ToString("O")
        });

        var transitioned = 0;
        var failed = 0;

        foreach (var invoice in invoices)
        {
            try
            {
                var didTransition = await ProcessInvoiceAsync(invoice, reference, cancellationToken);
                if (didTransition)
                    transitioned++;
            }
            catch (Exception ex)
            {
                failed++;
                logger.Error("Failed to transition overdue invoice", new Dictionary<string, object?>
                {
                    ["invoiceId"] = invoice.Id,
                    ["clientId"] = invoice.ClientId,
                    ["status"] = InvoiceStatusValues.ToValue(invoice.Status),
                    ["dueDate"] = invoice.DueDate.ToString("O"),
                    ["error"] = ex.Message
                });
            }
        }

        logger.Info("Overdue invoice processing completed", new Dictionary<string, object?>
        {
            ["transitioned"] = transitioned,
            ["failed"] = failed
        });

        return new ProcessOverdueInvoicesResult(transitioned, failed);
    }

    private async Task<bool> ProcessInvoiceAsync(Invoice invoice, DateTime referenceDate, CancellationToken cancellationToken)
    {
        if (!invoice.ShouldTransitionToFirstReminder(referenceDate))
            return false;

        var nextStatus = invoice.GetStatusAfterBecomingOverdue();
        await invoiceRepository.UpdateStatusAsync(invoice.Id, nextStatus, cancellationToken);

        logger.Info("Overdue invoice transitioned to first reminder", new Dictionary<string, object?>
        {
            ["invoiceId"] = invoice.Id,
            ["clientId"] = invoice.ClientId,
            ["previousStatus"] = InvoiceStatusValues.ToValue(invoice.Status),
            ["newStatus"] = InvoiceStatusValues.ToValue(nextStatus),
            ["dueDate"] = invoice.DueDate.ToString("O")
        });
        return true;
    }
}
